File: solutions/3981-jump-game-ix/jump-game-ix.c
typedef struct {
    int value;
    int left;
    int right;
} Item;

int* maxValue(const int* nums, int numsSize, int* returnSize) {
    int* ans = (int*)malloc(numsSize * sizeof(int));
    memset(ans, 0, numsSize * sizeof(int));

    Item* stack = (Item*)malloc(numsSize * sizeof(Item));
    int stackSize = 0;
    for (int i = 0; i < numsSize; i++) {
        Item curr = {nums[i], i, i};
        while (stackSize > 0 && stack[stackSize - 1].value > nums[i]) {
            Item top = stack[stackSize - 1];
            stackSize--;
            curr.value = (top.value > curr.value) ? top.value : curr.value;
            curr.left = top.left;
        }

        stack[stackSize++] = curr;
    }

    for (int i = 0; i < stackSize; i++) {
        for (int j = stack[i].left; j <= stack[i].right; j++) {
            ans[j] = stack[i].value;
        }
    }

    free(stack);
    *returnSize = numsSize;
    return ans;
}

File: solutions/3981-jump-game-ix/jump-game-ix.rs
impl Solution {
    pub fn max_value(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();

        let mut ans = vec![0; n];

        struct Item {
            value: i32,
            left: usize,
            right: usize,
        }

        let mut stack: Vec<Item> = Vec::new();

        for i in 0..n {
            let mut curr = Item {
                value: nums[i],
                left: i,
                right: i,
            };

            while let Some(top) = stack.last() {
                if top.value > nums[i] {
                    let top = stack.pop().unwrap();
                    curr.value = curr.value.max(top.value);
                    curr.left = top.left;
                } else {
                    break;
                }
            }

            stack.push(curr);
        }

        for i in 0..stack.len() {
            for j in stack[i].left..=stack[i].right {
                ans[j] = stack[i].value;
            }
        }

        ans
    }
}